TypeCobol.LanguageServer/Workspace.cs

```diff
@@ -951,6 +952,192 @@ private void FinalCompilationStepCompleted(object cUnit, ProgramClassEvent progr
             DocumentModifiedEvent?.Invoke(fileUri, new EventArgs());
         }
 
+        #region Data Layout
+        private const char SPACE = ' ';
+        private const string INDENT = "  "; // Indentation = 2 spaces
+        private const string UNDEFINED = "***";
+        private const string FILLER = "FILLER";
+        private const string GROUP = "GROUP";
+        private const string OCCURS = "OCCURS {0}";
+        private const string PIC = "PIC {0}";
+        private const string REDEFINES = "REDEFINES {0}";
+
+        /// <summary>
+        /// Get the Data Layout rows for a Program or a Copy (output = CSV)
+        /// </summary>
+        /// <param name="compilationUnit">Compilation unit resulting from parsing the Program/Copy</param>
+        /// <param name="separator">Separator to be used in the rows</param>
+        /// <returns></returns>
+        public string[] GetDataLayoutAsCSV(CompilationUnit compilationUnit, string separator)
+        {
+            var rows = new List<string>();
+            foreach (var dataLayoutNode in CollectDataLayoutNodes(compilationUnit))
+            {
+                var row = CreateRow(dataLayoutNode, separator);
+                if (row != null)
+                {
+                    rows.Add(row);
+                }
+            }
+
+            return rows.ToArray();
+
+            static string CreateRow(Tuple<int, DataDefinition> dataLayoutNode, string separator)
+            {
+                var nodeLevel = dataLayoutNode.Item1;
+                var dataDefinition = dataLayoutNode.Item2;
+                if (IsIgnored(dataDefinition))
+                {
+                    // Ignore this node
+                    return null;
+                }
+
+                var row = new StringBuilder();
+
+                //TODO manage slack bytes (property is dataDefinition.SlackBytes)
+
+                // Line number (starting at 1)
+                AppendToRow(dataDefinition.CodeElement.GetLineInMainSource() + 1);
+
+                /// Level number
+                AppendToRow(dataDefinition.CodeElement.LevelNumber);
+
+                // Name (preceded by an indent depending on the node level)
+                row.Append(SPACE, nodeLevel);
+                AppendToRow(GetName(dataDefinition));
+
+                // Declaration (Picture, Usage, REDEFINES, OCCURS, ...)
+                AppendDeclarationToRow(dataDefinition);
+
+                // Start/End/Length
+                var start = dataDefinition.StartPosition;
+                var length = dataDefinition.PhysicalLength;
+                AppendToRow(start);
+                AppendToRow(GetEnd(start, length));
+                AppendToRow(length);
+
+                return row.ToString();
+
+                static bool IsIgnored(DataDefinition dataDefinition)
```

Comment:
This could be done directly in CollectDataLayoutNodes and removed from here.

Reply:
Yes indeed. Done